### PR TITLE
protocol: streamline asset ID definition

### PR DESCRIPTION
**Pull request by oleganza**
This change does two things:

1. **Removes asset version from Asset ID** to allow compatibility of the AssetIDs with future output types in the future protocol versions. Security of non-upgraded clients should not be affected as they already verify balancing rule, therefore, preventing unaccounted transfer of assets from the old entries to the new ones.
2. **Moves asset definition from the issuance program into its own field in the issuance input**. So we no longer have clumsy `<data> DROP <actual code>` issuance programs, have an explicit asset definition field in the issuance input and, as a result, support serialization flags for asset definition just like for the rest of the reference data.

This is a part of a package of breaking changes in P1: https://github.com/chain/chain/issues/239

### CODE REVIEW

File: docs/protocol/specifications/data.md [hunk 1]
```diff
@@ -309,6 +309,7 @@ VM Version              | varint63                | [Version of the VM](#vm-vers
 Issuance Program        | varstring31             | Predicate defining the conditions of issue.
 Program Arguments Count | varint31                | Number of [program arguments](#program-arguments) that follow.
 Program Arguments       | [varstring31]           | [Signatures](#signature) and other data satisfying the spent output’s predicate. Used to initialize the [data stack](vm1.md#vm-state) of the VM that executes an issuance or an control program.
+Asset Definition        | varstring31             | Arbitrary string or its [optional hash](#optional-hash), depending on [serialization flags](#transaction-serialization-flags).
```
Comment (oleganza) [inline docs/protocol/specifications/data.md]: I'd suggest doing that as a separate PR and changes in all the docs etc.

**Comment by kr** [comment]
LGTM

**Comment by oleganza** [comment]
@kr @danrobinson PTAL. Implementation is fully done. 

WARNING: this is a hard-forking change to the protocol.

**Comment by oleganza** [comment]
Note to self: rename `EmptyHash` to `EmptyStringHash`.

**Comment by kr** [comment]
:eyes:

File: protocol/block_test.go [hunk 2]
```diff
@@ -97,7 +97,7 @@ func TestWaitForBlockSoonWaits(t *testing.T) {
 	makeEmptyBlock(t, c) // height=2
 
 	go func() {
-		time.Sleep(10 * time.Millisecond) // sorry for the slow test 😔
+		time.Sleep(10 * time.Millisecond) // sorry for the slow test 
```
Comment (kr) [inline protocol/block_test.go]: huh, was this a deliberate change?

File: protocol/bc/asset.go [hunk 3]
```diff
@@ -22,13 +22,17 @@ func (a *AssetID) Scan(b interface{}) error     { return (*Hash)(a).Scan(b) }
...
 // the given issuance program and initial block hash.
-func ComputeAssetID(issuanceProgram []byte, initialHash [32]byte, vmVersion uint64) (assetID AssetID) {
+func ComputeAssetID(issuanceProgram []byte, initialHash [32]byte, vmVersion uint64, assetDefinitionHash Hash) (assetID AssetID) {
 	h := sha3pool.Get256()
 	defer sha3pool.Put256(h)
 	h.Write(initialHash[:])
-	blockchain.WriteVarint63(h, assetVersion)
 	blockchain.WriteVarint63(h, vmVersion)
 	blockchain.WriteVarstr31(h, issuanceProgram) // TODO(bobg): check and return error
+	if assetDefinitionHash == EmptyStringHash {
```
Comment (kr) [inline protocol/bc/asset.go]: Why check for this? It seems like it would be easier to just hash the asset definition hash regardless. Presumably there's a drawback to that but I don't understand what it is.

**Comment by kr** [comment]
(I edited the PR title to change to imperative from past tense)

File: protocol/block_test.go (hunk 2 above)
Comment (oleganza) [inline protocol/block_test.go]: oops, maybe my textmate butchered that emoji.

Comment (oleganza) [inline protocol/block_test.go]: Fixed in https://github.com/chain/chain/pull/261/commits/8283533fcc9a0404c2bf4f8101a1777b95e50028

File: protocol/bc/asset.go (hunk 3 above)
Comment (oleganza) [inline protocol/bc/asset.go]: Replaced optional hash here with normal hash. Did not touch optional hashes in tx structures yet - will do in a separate PR. Cf. https://github.com/chain/chain/pull/261/commits/6d6ccb1cfe245b48032211a14fa1f1c82d0421c5

**Comment by kr** [comment]
LGTM!
